src/Files.App.Controls/ThemedIcon/ThemedIcon.cs

```diff
@@ -6,371 +6,224 @@
-            // Then we check if the Disabled Icon is Toggled
-
-            // We then assume the Icon is Enabled
-            // We then check the Toggled state for the Contrast Icons
-            // We have two states depending on toggle.
-
-            // Finally we act on all other Enabled states
-            // We check for Toggled state
-            // And update the IconColorType in the Layered Icon's Layers
-            */
-
-            if (_isEnabled is false || IsEnabled is false)
-            {
-                if (_isToggled is true || IsToggled is true)
-                {
-                    VisualStateManager.GoToState(this, DisabledToggleColorStateName, true);
-                }
-                else
-                {
-                    VisualStateManager.GoToState(this, DisabledColorStateName, true);
-                }
-            }
-            else
-            {
-                if (_isToggled is true || IsToggled is true)
-                {
-                    VisualStateManager.GoToState(this, ToggleStateName, true);
-                }
-                else
-                {
-                    VisualStateManager.GoToState(
-                        this,
-                        IconColorType switch
-                        {
-                            ThemedIconColorType.Critical => CriticalStateName,
-                            ThemedIconColorType.Caution => CautionStateName,
-                            ThemedIconColorType.Success => SuccessStateName,
-                            ThemedIconColorType.Neutral => NeutralStateName,
-                            ThemedIconColorType.Accent => AccentStateName,
-                            ThemedIconColorType.Custom => CustomColorStateName,
-                            _ => NormalStateName,
-                        },
-                        true);
-                }
-
-                if (GetTemplateChild(LayeredPathCanvas) is Canvas canvas)
-                {
-                    foreach (var layer in canvas.Children.Cast<ThemedIconLayer>())
-                        layer.IconColorType = IconColorType;
-                }
-            }
-        }
-    }
+	/// <summary>
+	/// A control for a State and Color aware Icon
+	/// </summary>
+	public partial class ThemedIcon : Control
+	{
+		public ThemedIcon()
+		{
+			DefaultStyleKey = typeof(ThemedIcon);
+		}
+
+		protected override void OnApplyTemplate()
+		{
+			IsEnabledChanged -= OnIsEnabledChanged;
+
+			base.OnApplyTemplate();
+
+			IsEnabledChanged += OnIsEnabledChanged;
+
+			_isOwnerEnabled = IsEnabled;
+
+			FindOwnerControlStates();
+			OnFilledIconChanged();
+			OnOutlineIconChanged();
+			OnLayeredIconChanged();
+
+			OnIconTypeChanged();
+			OnIconColorTypeChanged();
+		}
+
+		// Updates paths and layers
+
+		private void OnFilledIconChanged()
+		{
+			// Updates Filled Icon from Path Data
+			if (GetTemplateChild(FilledPathIconViewBox) is not Viewbox filledViewBox)
```

Comment:
I'm curious how often we are querying the Visual Tree (not only here but in general in this file) and if there might be a benefit in having a reference to these parts. That's what most of the WinUI controls do too.

Reply:
Do you think we should query them one time when applying template?

Reply:
Yes, that's what WinUI does too for performance reason.

Reply:
I have added a method to get the template parts and set the private fields - and then we check the if those fields are null in the code.